agent/lib/kontena/observer.rb

```diff
@@ -89,29 +268,103 @@ def update_observe(observe, observable, value)
     # @raise failed to observe observables
     # @return [Observer::Observe]
     # @yield [*values] all Observables are ready
-    def observe(*observables, &block)
+    def observe_async(*observables, &block)
+      self.register_observer_handler
+
+      actor = Celluloid.current_actor
+
       # unique handle to identify this observe loop
-      observe = Observe.new(self.class.name, observables, block)
+      observe = Observe::Async.new(self.class, &block)
 
       # sync setup of each observable
       observables.each do |observable|
-        # register for async.update_observe(...)
-        value = observable.add_observer(Celluloid.current_actor, observe)
+        # register for observable updates
+        # this MUST be atomic with the Observe#add, there cannot be any observe update message in between!
+        message = observable.add_observer(actor.mailbox, observe)
 
-        # store value for initial call, or nil to block
-        observe.set(observable, value)
+        if message.value
+          debug "observe async #{message.describe_observable} => #{message.value}"
 
-        debug "observe Observable<#{observable.__klass__}> -> #{value}"
+          observe.add(message.observable, message.value)
+        else
+          debug "observe async #{message.describe_observable}..."
+
+          observe.add(message.observable)
+        end
 
         # crash if observed Actor crashes, otherwise we get stuck without updates
         # this is not a bidrectional link: our crashes do not propagate to the observable
         self.monitor observable
```

Comment:
This won't work for standalone Observables... the sync observe is also missing linking to notice crashing observables, although it does have the timeout support.